File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None
        map = {}
        q = deque([node])

        while q:
            cur = q.popleft()
            if map.get(cur) is None:
                map[cur] = Node(cur.val)
            
            for nei in cur.neighbors:
                if map.get(nei) is None:
                    map[nei] = Node(nei.val)
                    q.append(nei)
                    
                map[nei].neighbors.append(map.get(cur))

        return map.get(node)

File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return node;
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> q = new ArrayDeque<>();

        q.add(node);
        while (!q.isEmpty()) {
            Node cur = q.poll();
            if (map.get(cur) == null) {
                map.put(cur, new Node(cur.val));
            }

            for (Node nei : cur.neighbors) {
                if (map.get(nei) == null) {
                    map.put(nei, new Node(nei.val));
                    q.add(nei);
                }
                map.get(cur).neighbors.add(map.get(nei));
            }
        }

        return map.get(node);
    }
}